### PR TITLE
Add metric for calculating index flush time excluding waiting on locks

### PR DESCRIPTION
**Pull request by arteam**
Add a new `total_time_excluding_waiting_on_lock` metric to index flush stats that measures the flushing time excluding waiting on the flush lock. This metrics provides a more granular view on flush performance and without the overhead of flush throttling.

Resolves ES-7201

### CODE REVIEW

**Comment by arteam** [comment]
@elasticmachine update branch

**Comment by arteam** [comment]
@elasticmachine update branch

File: server/src/main/java/org/elasticsearch/index/flush/FlushStats.java [hunk 1]
```diff
@@ -32,18 +34,21 @@ public FlushStats(StreamInput in) throws IOException {
...
         this.totalTimeInMillis = totalTimeInMillis;
+        this.totalTimeExcludingWaitingInMillis = totalTimeExcludingWaitingInMillis;
     }
 
-    public void add(long total, long periodic, long totalTimeInMillis) {
+    public void add(long total, long periodic, long totalTimeInMillis, long totalTimeWithoutWaitingInMillis) {
         this.total += total;
         this.periodic += periodic;
         this.totalTimeInMillis += totalTimeInMillis;
+        this.totalTimeExcludingWaitingInMillis = totalTimeWithoutWaitingInMillis;
```
Comment (fcofdez) [inline server/src/main/java/org/elasticsearch/index/flush/FlushStats.java:51]: we should add `totalTimeWithoutWaitingInMillis` here too I think?

File: server/src/main/java/org/elasticsearch/index/engine/InternalEngine.java [hunk 2]
```diff
@@ -3066,6 +3072,11 @@ long getNumDocUpdates() {
         return numDocUpdates.count();
     }
 
+    @Override
+    public long getFlushTimeExcludingWaitingInMillis() {
```
Comment (fcofdez) [inline server/src/main/java/org/elasticsearch/index/engine/InternalEngine.java:3076]: this is the total time spent on flush, maybe we should rename the field?

**Review by fcofdez — COMMENTED** [review]
Looks good, I think we should expose this through APM and maybe leave out the cat API changes? wdyt?

File: server/src/main/java/org/elasticsearch/index/flush/FlushStats.java [hunk 3]
```diff
@@ -23,6 +24,7 @@ public class FlushStats implements Writeable, ToXContentFragment {
     private long total;
     private long periodic;
     private long totalTimeInMillis;
+    private long totalTimeExcludingWaitingInMillis;
```
Comment (volodk85) [inline server/src/main/java/org/elasticsearch/index/flush/FlushStats.java:27]: without *waiting* for what exactly? is it for commit durability? I miss a bit better naming here. wdyt?

**Review by fcofdez — COMMENTED** [review]
Looks good, I left a couple of minor comments before approving

File: server/src/main/java/org/elasticsearch/index/engine/InternalEngine.java (hunk 2 above)
Comment (arteam) [reply server/src/main/java/org/elasticsearch/index/engine/InternalEngine.java]: Good point, I've changed it :+1:

File: server/src/test/java/org/elasticsearch/index/shard/IndexShardTests.java [hunk 4]
```diff
@@ -3974,6 +3977,39 @@ public void testFlushOnIdle() throws Exception {
...
+        for (int i = 0; i < randomIntBetween(4, 10); i++) {
+            indexDoc(shard, "_doc", Integer.toString(i));
+        }
+
+        int numFlushes = randomIntBetween(2, 5);
+        var flushesLatch = new CountDownLatch(numFlushes);
+        var executor = Executors.newFixedThreadPool(numFlushes);
+        for (int i = 0; i < numFlushes; i++) {
+            executor.submit(() -> {
+                shard.flush(new FlushRequest().waitIfOngoing(true));
```
[the remark this answers is not in the document]
Comment (arteam) [reply server/src/test/java/org/elasticsearch/index/shard/IndexShardTests.java]: Good catch, I've added `.force(true)` call :+1:

File: server/src/main/java/org/elasticsearch/TransportVersions.java [hunk 5]
```diff
@@ -171,6 +171,7 @@ static TransportVersion def(int id) {
     public static final TransportVersion MODIFY_DATA_STREAM_FAILURE_STORES = def(8_630_00_0);
     public static final TransportVersion ML_INFERENCE_RERANK_NEW_RESPONSE_FORMAT = def(8_631_00_0);
     public static final TransportVersion HIGHLIGHTERS_TAGS_ON_FIELD_LEVEL = def(8_632_00_0);
+    public static final TransportVersion ACTUAL_FLUSH_STATS = def(8_633_00_0);
```
[the remark this answers is not in the document]
Comment (arteam) [reply server/src/main/java/org/elasticsearch/TransportVersions.java]: Absolutely! I've changed it to `TRACK_FLUSH_TIME_EXCLUDING_WAITING_ON_LOCKS`

**Comment by arteam** [comment]
@elasticmachine update branch

**Review by fcofdez — APPROVED** [review]
LGTM 👍

**Comment by arteam** [comment]
Thank you!

**Pull request merged**